### PR TITLE
PSP-7509 : Allow Contractor to Create or Edit Disposition File

### PR DESCRIPTION
**Pull request by eddherrera**

### CODE REVIEW

File: source/backend/api/Helpers/Extensions/DispositionFileExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            ArgumentNullException.ThrowIfNull(dispositionFile);
+
+            ArgumentNullException.ThrowIfNull(principal);
+
+            var pimsUser = userRepository.GetUserInfoByKeycloakUserId(principal.GetUserKey());
+
+            if (pimsUser?.IsContractor == true && !dispositionFile.PimsDispositionFileTeams.Any(x => x.PersonId == pimsUser.PersonId))
+            {
+                throw new UserOverrideException(UserOverrideCode.ContractorNotInTeam, "As a Contractor your user contact information should be assigned to the Disposition File's team");
```
Comment (devinleighsmith) [inline source/backend/api/Helpers/Extensions/DispositionFileExtensions.cs:23]: This doesn't seem to match the error text provided in the story:
As a contractor, you must add yourself as a team member to the file in order to create or save changes.

Also, the story specifies that this should be thrown as an error. The contract user should not be allowed to override this error - the system should completely disallow saving a file when a contractor is creating it and not on the team.

UserOverrideExceptions should only be used when the user is being shown a warning that can be optionally overriden.

File: source/backend/api/Services/DispositionFileService.cs [hunk 2]
```diff
@@ -52,12 +55,14 @@ public class DispositionFileService : IDispositionFileService
...
             _lookupRepository = lookupRepository;
             _checklistRepository = checklistRepository;
+            _userRepository = userRepository;
         }
 
         public PimsDispositionFile Add(PimsDispositionFile dispositionFile, IEnumerable<UserOverrideCode> userOverrides)
         {
             _logger.LogInformation("Creating Disposition File {dispositionFile}", dispositionFile);
             _user.ThrowIfNotAuthorized(Permissions.DispositionAdd);
+            dispositionFile.ThrowMissingContractorInTeam(_user, _userRepository);
```
Comment (devinleighsmith) [inline source/backend/api/Services/DispositionFileService.cs:65]: is this function logically equivalent to: _user.ThrowInvalidAccessToDispositionFile(_userRepository, _dispositionFileRepository, id);
?
If so I'd rather not have two functions that provide the same functionality. Especially given that this function seems to require that the disposition team be provided with the file which is not a guarantee.

File: source/frontend/src/models/api/UserOverrideCode.ts [hunk 3]
```diff
@@ -3,6 +3,7 @@ export enum UserOverrideCode {
   ADD_LOCATION_TO_PROPERTY = 'ADD_LOCATION_TO_PROPERTY',
   UPDATE_REGION = 'UPDATE_REGION',
   PROPERTY_OF_INTEREST_TO_INVENTORY = 'PROPERTY_OF_INTEREST_TO_INVENTORY',
+  CONTRACTOR_NOT_IN_TEAM = 'CONTRACTOR_NOT_IN_TEAM',
```
Comment (devinleighsmith) [inline source/frontend/src/models/api/UserOverrideCode.ts:6]: Both of the below two override codes seem suspect, as a user should not be able to override not being in the team or removing themselves.

File: source/backend/api/Services/DispositionFileService.cs [hunk 4]
```diff
@@ -88,6 +94,7 @@ public PimsDispositionFile Update(long id, PimsDispositionFile dispositionFile,
 
             _logger.LogInformation("Updating acquisition file with id {id}", id);
             _user.ThrowIfNotAuthorized(Permissions.DispositionEdit);
+            _user.ThrowInvalidAccessToDispositionFile(_userRepository, _dispositionFileRepository, id);
```
Comment (devinleighsmith) [inline source/backend/api/Services/DispositionFileService.cs:97]: Hmm, doesn't this one also need to know if the incoming disposition file is removing the current contractor from the disposition team.

File: source/backend/api/Helpers/Extensions/PrincipalExtensions.cs [hunk 5]
```diff
@@ -11,18 +11,28 @@ public static class PrincipalExtensions
...
+        public static void ThrowInvalidAccessToDispositionFile(this ClaimsPrincipal principal, IUserRepository userRepository, IDispositionFileRepository dispositionFileRepository, long dispostionFileId)
+        {
+            ArgumentNullException.ThrowIfNull(principal);
+
+            var pimsUser = userRepository.GetUserInfoByKeycloakUserId(principal.GetUserKey());
+            PimsDispositionFile dispostionFile = dispositionFileRepository.GetById(dispostionFileId);
+
+            if (pimsUser?.IsContractor == true && !dispostionFile.PimsDispositionFileTeams.Any(x => x.PersonId == pimsUser.PersonId))
+            {
+                throw new NotAuthorizedException("Contractor is not assigned to the Disposition File's team");
```
Comment (devinleighsmith) [inline source/backend/api/Helpers/Extensions/PrincipalExtensions.cs:34]: This seems better, if a contractor does not have access to a file they should see an error not an override.

File: source/frontend/src/hooks/useApiUserOverride.ts [hunk 6]
```diff
@@ -70,59 +70,75 @@ export const useApiUserOverride = <
...
-        title: 'Note',
-        variant: 'info',
-        message: RemoveSelfContractorContent(),
-        handleOk: async () => {
-          setState({
-            previousUserOverrideCodes: [...state.previousUserOverrideCodes],
-            userOverrideCode: null,
-            message: null,
+          break;
+        case UserOverrideCode.CONTRACTOR_SELFREMOVED:
```
Comment (devinleighsmith) [inline source/frontend/src/hooks/useApiUserOverride.ts:92]: Yes, the fact that the userOverrideCode is null here is a red flag. Why are we using the user override logic when the user isn't able to override anything?

File: source/frontend/src/hooks/useApiUserOverride.ts [hunk 7]
```diff
@@ -70,59 +70,75 @@ export const useApiUserOverride = <
...
+        case UserOverrideCode.CONTRACTOR_NOT_IN_TEAM:
+          setModalContent({
+            title: 'Contractor missing',
+            variant: 'warning',
+            message:
+              'As a contractor, you must add yourself as a team member to the file in order to create or save changes.',
+            handleOk: async () => {
+              setState({
+                previousUserOverrideCodes: [...state.previousUserOverrideCodes],
+                userOverrideCode: null,
```
Comment (devinleighsmith) [inline source/frontend/src/hooks/useApiUserOverride.ts:84]: same problem here as below.

File: source/frontend/src/hooks/useApiUserOverride.ts (hunk 6 above)
Comment (devinleighsmith) [reply source/frontend/src/hooks/useApiUserOverride.ts:92]: > Yes, the fact that the userOverrideCode is null here is a red flag. Why are we using the user override logic when the user isn't able to override anything?

This code demonstrates the correct way to handle network errors (non 409 responses) in concert with the UserOverride logic:

`withUserOverride(
          (userOverrideCodes: UserOverrideCode[]) =>
            handleSubmit(values, formikHelpers, userOverrideCodes),
          [],
          (axiosError: AxiosError<IApiError>) => { // this function is executed when a non-409 400 error code is returned.
            setModalContent({
              variant: 'warning',
              title: 'Warning',
              message: axiosError?.response?.data.error,
              okButtonText: 'Close',
            });
            setDisplayModal(true);
          },
        )`

File: source/backend/api/Services/DispositionFileService.cs (hunk 4 above)
Comment (eddherrera) [reply source/backend/api/Services/DispositionFileService.cs:97]: that's another story.

File: source/backend/api/Helpers/Extensions/DispositionFileExtensions.cs (hunk 1 above)
Comment (eddherrera) [reply source/backend/api/Helpers/Extensions/DispositionFileExtensions.cs]: Updated

File: source/frontend/src/models/api/UserOverrideCode.ts (hunk 3 above)
Comment (eddherrera) [reply source/frontend/src/models/api/UserOverrideCode.ts]: Removed

File: source/backend/api/Helpers/Extensions/PrincipalExtensions.cs (hunk 5 above)
Comment (eddherrera) [reply source/backend/api/Helpers/Extensions/PrincipalExtensions.cs:34]: all good

File: source/frontend/src/hooks/useApiUserOverride.ts (hunk 6 above)
Comment (eddherrera) [reply source/frontend/src/hooks/useApiUserOverride.ts:75]: Updated

File: source/frontend/src/hooks/useApiUserOverride.ts (hunk 7 above)
Comment (eddherrera) [reply source/frontend/src/hooks/useApiUserOverride.ts]: Updated

File: source/backend/api/Helpers/Extensions/PrincipalExtensions.cs [hunk 8]
```diff
@@ -11,18 +11,28 @@ public static class PrincipalExtensions
...
                 throw new NotAuthorizedException("Contractor is not assigned to the Acquisition File's team");
             }
         }
+
+        public static void ThrowInvalidAccessToDispositionFile(this ClaimsPrincipal principal, IUserRepository userRepository, IDispositionFileRepository dispositionFileRepository, long dispostionFileId)
+        {
+            ArgumentNullException.ThrowIfNull(principal);
+
+            var pimsUser = userRepository.GetUserInfoByKeycloakUserId(principal.GetUserKey());
+            PimsDispositionFile dispostionFile = dispositionFileRepository.GetById(dispostionFileId);
```
Comment (devinleighsmith) [inline source/backend/api/Helpers/Extensions/PrincipalExtensions.cs:30]: nit: dispostion

**Review by devinleighsmith — APPROVED** [review]

File: source/backend/api/Helpers/Extensions/PrincipalExtensions.cs (hunk 8 above)
Comment (eddherrera) [reply source/backend/api/Helpers/Extensions/PrincipalExtensions.cs:30]: Fixing on next pr.

**Pull request merged**